### PR TITLE
Lerdge S,X,K Board Support

### PR DESCRIPTION
**Pull request by xC0000005**
### Requirements

* Filling out this template is required. Pull Requests without a clear description may be closed at the maintainers' discretion.

### Description

This provides the ability to build and run Marlin 2.0 on Lerdge X board. The firmware is automatically encrypted to support SD based update. LCD Support does not work yet (I do not know why - I've tried with @jmz52's PR and it doesn't initialize yet (probably something I configured wrong).

### Benefits

If you were not me, running Marlin on Lerdge wasn't easy. Now anyone can do it.
Lerdge K and K-2 support will require a different variant but use the same encryption scripts.

### Related Issues

#12624

### CODE REVIEW

File: Marlin/src/pins/stm32f4/pins_LERDGE_X.h
Comment (thinkyhead) [inline Marlin/src/pins/stm32f4/pins_LERDGE_X.h:146]: Marlin doesn't know these pin names, so I assume the previous pins names were correct.

**Comment by thinkyhead** [comment]
> I've tried with @jmz52's PR and it doesn't initialize yet

Now that the FSMC pin names are corrected, it might work….

**Comment by xC0000005** [comment]
> > I've tried with @jmz52's PR and it doesn't initialize yet
> 
> Now that the FSMC pin names are corrected, it might work….

The LCD requires a different set of massaging sequences before it will behave. I have it working in another branch - once the TFT support PR is merged, we can update with the Lerdge Init sequence and it'll run (see my comments on the TFT PR for a picture of Marlin 2.x goodness).

The biggest set of outstanding work is figuring out how to generalize the variant to support Lerdge X, K and S (if that can be done). Otherwise, I'll work on K support next since I have one.

**Comment by xC0000005** [comment]
Tested after your updates, it's still booting and responding, so that's good.

**Comment by jmz52** [comment]
There is a shared variant for black F407Vx and F407Zx boards - BLACK_F407XX
It uses 4 defines to tweak configuration.
```
ARDUINO_BLACK_F407VE
ARDUINO_BLACK_F407VG
ARDUINO_BLACK_F407ZE
ARDUINO_BLACK_F407ZG
```

**Comment by xC0000005** [comment]
My Lerdge K board is dead (I was warned it might be, the goal was to dump the bootloader, and we did succeed at that), so I won't be adding K or S variant support for right now.

**Comment by thinkyhead** [comment]
Does this mean that the LERDGE_X variant included here could be removed and replaced with `variant = BLACK_F407XX` and one of the 4 flags to specify the exact model? Or, does that comment only apply the the LERDGE K variant?

**Comment by jmz52** [comment]
BLACK_F407XX is an example how boards with different MCUs can share a single 'variant' definition.
Something similar might be used to create variant for all three lerdge boards.

**Comment by xC0000005** [comment]
That’s the eventual Idea. Since my K isn’t running right now, I can’t test it. But at some point a basic “Lerdge” variant could support all. At some point I might be able to make it work with an existing board variant.

**Comment by xC0000005** [comment]
After checking it over, the Ledge variant uses different clock config than the black, but it is near certain that the same "Lerdge" variant could be used across S, X, and K.

File: buildroot/share/PlatformIO/ldscripts/lerdge.ld
Comment (xC0000005) [inline buildroot/share/PlatformIO/ldscripts/lerdge.ld]: This is not correct - I'll push an update for it. The board has 512k minus 64 for the boot loader.

**Comment by thinkyhead** [comment]
I assume we are still waiting on #18130 before merging this (and vice-versa)…?

**Comment by xC0000005** [comment]
We can wait on this - I got confirmation from someone that Marlin runs on the Ledge-S (no LCD and wrong pins), and we know the same variant runs on the K (but needs more pins defined), so I’ll rename it to LERDGE rather than LERDGE_X. After that, if all the tests pass, this part can be merged, and I’ll continue to figure out the issues with the LCD next.

**Comment by mapfde** [comment]
Here is my Lerdge-S board running Marlin.

**Comment by jmz52** [comment]
If I recall correctly the `VECT_TAB_ADDR` define was used in old framework-arduinoststm32-maple framework only. It has no effect on framework-arduinoststm32.
New framework-arduinoststm32 framework and HAL STM32 use `VECT_TAB_OFFSET` define.

**Comment by xC0000005** [comment]
You are correct - this is left over from me trying to get this building with Maple. I’ll remove it.

**Comment by xC0000005** [comment]
Confirmed that Marlin also works in a very limited fashion on the K (more a test of whether the variant can be expanded to work with it than a real build yet).

**Comment by mapfde** [comment]
Finished mapping the pins to Lerdge-S. See https://github.com/xC0000005/Marlin/pull/10

**Comment by xC0000005** [comment]
Since we have S pins, we can hold merging this and I will rework it into full s/x/k support.

**Comment by xC0000005** [comment]
I made a mistake pulling in the S commits, Ignore this for now, I'll figure out how to get rid of accidentally pulling in the TFT PR.

**Comment by thinkyhead** [comment]
If you feel like rebuilding your commits…
```
git fetch origin
git fetch upstream
git reset --hard origin/LerdgeXBasicSupport
git reset --soft upstream/bugfix-2.0.x
```
This will leave you with `bugfix-2.0.x` with all changes, uncommitted. You can then make all new commits and use `git push -f` to update the PR.

**Comment by Evg33** [comment]
xC0000005: Thank you !!!

I compiled LCDPR branch yesterday.

I got an error and fixed it for a successful compilation for me.
\buildroot\share\PlatformIO\scripts\lerdge.py 
```
  #print("You need to define output file via '-DFIRMWARE' build parameter", file=sys.stderr)
  print("You need to define output file via '-DFIRMWARE' build parameter")
```

**Comment by xC0000005** [comment]
The compile error should not happen in the most recent PR (but that doesn’t have the LCD working yet). If you hardcode the ledge init sequence and roll back to around commit 2d8f324b9ad5eb93a03bdab3ab3d2d151c8ba3ee, the LCD should work with correct colors (make sure to change the buffer size in TFT.h to something smaller - whatever the 103 is using). Next week I’ll run down what commit broke them, because even switching to the Ledge init sequence doesn’t fix it in the most recent code of that PR. We’ll get there. I need to put a multimeter on and find out what pin is used to enable the max for thermocouple as well, and how the power management module works, but S support is good, X support is pretty good, K is very iffy and needs work. I do not expect SPI TMC drivers to work with the K yet.

**Comment by thisiskeithb** [comment]
From Discord user: mmpp0

> You just changed the Lerdge pins files, which is fine. But the comment became very misleading. Please change:

> `// Thermistor pins must be OUT/LOW. Done automatically for thermocouples.`

> to

> `// For thermistors the following pins must be OUT/LOW. Done automatically according to the configured TEMP_SENSOR_0 and TEMP_SENSOR_1.`

File: Marlin/src/pins/stm32f4/pins_LERDGE_X.h
Comment (thinkyhead) [inline Marlin/src/pins/stm32f4/pins_LERDGE_X.h]: Random user asks: "What do I need to do to make these pins OUT/LOW?"

Comment (mapfde) [reply Marlin/src/pins/stm32f4/pins_LERDGE_X.h]: Like I wrote in the update to the comment in the pins file: It is done automatically now, according to the configured thermistor. The code is in the temperature class initialization.

One can also use M42 to toggle the pin manually for testing purposes.

Comment (mapfde) [reply Marlin/src/pins/stm32f4/pins_LERDGE_X.h]: I just noticed that the question was about Lerdge-X - this board does not support thermocouples.

**Comment by xC0000005** [comment]
Correct - in my reverse engineered documentation, I have a pin labeled for the MAX, but the board doesn’t have it (and I have no idea what I was thinking). I have some fixes to the K pins as well, now that I have Marlin running reliably on the K (without LCD). I’ll make a few adjustments tonight after I submit the changes to the LCD PR.

**Comment by Evg33** [comment]
Thermocouples only with Lerdge-K. 
Lerdge-X not support TC. only NTC100.

**Comment by xC0000005** [comment]
Correct. I will fix that tonight. I am also working on pins updates for the lerdge k, since it has a number of issues.

**Comment by xC0000005** [comment]
@thinkyhead , is there an astyle file I can run against changes before pushing them to make sure formatting matches standards?

**Comment by thinkyhead** [comment]
> there an astyle file I can run against changes

There's a script `buildroot/bin/uncrust` that runs `uncrustify` on the given pathname using rules in `buildroot/share/extras` that are pretty close to our existing formatting. The current version of `uncrustify` is a little bit off, and it doesn't do a perfect job with #preproc indentation. So some tuning is usually required afterward.

There is also a script `buildroot/share/scripts/pinsformat.js` script which specifically applies formatting to `pins/pins_MY_BOARD.h` files. For example…

```sh
find Marlin/src/pins -name pins_*.h -exec ./buildroot/share/scripts/pinsformat.js '{}' \;
```

**Comment by cjsoong** [comment]
@xC0000005 
I compiled LCDPR branch yesterday.
TFT display setting:
#define TFT_480x320 
#define TFT_DRIVER LERDGE_ST7796
 bug report:
I found some problems, Y-min is always in the triggered state.
The rotary switch on the TFT cannot be used, and the value keeps jumping.

**Comment by xC0000005** [comment]
For what board?

**Comment by cjsoong** [comment]
@xC0000005 
Lerdge-X

**Comment by xC0000005** [comment]
Can you share with me your config? My endstops behave, as does the encoder. If I set the board to LERDGE_S in configuration.h, I get similar behavior.

**Comment by cjsoong** [comment]
@xC0000005 
[LerdgeX_config.zip](https://github.com/MarlinFirmware/Marlin/files/4849601/LerdgeX_config.zip)

**Comment by xC0000005** [comment]
Thank you. I just confirmed that with the default config (board set to x, tft and driver, serial, baud), Rotary works smooth and fine, and all endstops function as expected. I’ll compile with yours tomorrow and see what I see.

**Comment by xC0000005** [comment]
On my boards, with your config, I cannot repro this.  Endstops show the correct value open or closed, The encoder is smooth as well. I moved it over to the Lerdge-K and tested there, and endstops report the correct values (as in, change if open or closed as appropriate). I tested X, Y and Z, and the look right.  Touch calibration sometimes fails, but I’m looking into that.

Which values are jumping for you?

**Comment by jmz52** [comment]
Touch calibration code compares raw data for points with same coordinate.
If there are more that 20% difference between reading the calibration will fail. I am using stylus for calibration and typical deviation is less than 10%.
Without stylus 20% deviation limit might be too strict.
This can be adjusted by changing parameter in touch.h
`#define TOUCH_CALIBRATION_PRECISION 80`

**Comment by xC0000005** [comment]
Also, there’s a bug in my hold to calibrate that can make the screen loop. I will figure it out and fix it later. Do we plan to store calibration data in EEPROM (or is it already done?)

**Comment by jmz52** [comment]
Already done.
As for hold to calibrate - I like the idea, it is easier that calling M995 from console.
The only change I would make it limit "hold to calibrate" functionality to status screen only to avoid conflict with slider (and to be able to use "hold" interactions in other scenarios, UBL for example)

**Comment by thinkyhead** [comment]
Checking in… How is this looking to the Lerdge experts?

**Comment by Evg33** [comment]
I change 1 line in lerdge.py for success compilation with Python 2.7.16.
https://github.com/xC0000005/Marlin/blob/2467fa195b7d133e026126f722b105779b81c698/buildroot/share/PlatformIO/scripts/lerdge.py#L45
```
  #print("You need to define output file via '-DFIRMWARE' build parameter", file=sys.stderr)
  print("You need to define output file via '-DFIRMWARE' build parameter")
```

without it:
```
> pio run
Processing LERDGEX (platform: ststm32@<6.2.0; board: LERDGE; framework: arduino)
--------------------------------------------------------------------------------------------------------------------------------- 
Verbose mode can be enabled via `-v, --verbose` option
('Starting SRC Filter:', ['+<src/*> -<src/config> -<src/HAL> +<src/HAL/shared> +<src/HAL/STM32>'])
('Modified SRC Filter:', ['+<src/*> -<src/config> -<src/HAL> +<src/HAL/shared> +<src/HAL/STM32> +<buildroot/share/PlatformIO/variants/LERDGE>'])
('CXXFLAGS', [])
CONFIGURATION: https://docs.platformio.org/page/boards/ststm32/LERDGE.html
PLATFORM: ST STM32 6.1.1 > STM32F407ZGT6(192k RAM. 1024k Flash)
HARDWARE: STM32F407ZGT6 168MHz, 192KB RAM, 1MB Flash
DEBUG: Current (jlink) External (jlink, stlink)
PACKAGES:
 - framework-arduinoststm32 4.10700.200103 (1.7.0)
 - toolchain-gccarmnoneeabi 1.90201.191206 (9.2.1)
Converting Marlin.ino
Warning! Cannot find linker script for the current target!

LDF: Library Dependency Finder -> http://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 19 compatible libraries
Scanning dependencies...
Dependency Graph
|-- <LiquidCrystal> 1.5.0
|   |-- <Wire> 1.0
|-- <TMCStepper> 0.7.0
|   |-- <SPI> 1.0
|-- <Adafruit MAX31865 library> 1.1.0
|   |-- <SPI> 1.0
|-- <Adafruit NeoPixel> 1.5.0
|-- <U8glib-HAL> 0.4.1
|   |-- <Wire> 1.0
|   |-- <SPI> 1.0
|-- <Arduino-L6470> 0.8.0
|-- <SlowSoftI2CMaster>
|-- <LiquidTWI2> 1.2.7
|   |-- <Wire> 1.0
|-- <SailfishLCD>
|-- <Wire> 1.0
|-- <Servo> 1.1.2
|-- <IWatchdog> 1.0.0
|-- <SPI> 1.0
Building in release mode
  File "X:\Marlin\buildroot\share\PlatformIO\scripts\lerdge.py", line 45

    print("You need to define output file via board_build.firmware = 'filename' parameter", file=sys.stderr)

                                                                                                ^

SyntaxError: invalid syntax

================================================== [FAILED] Took 5.55 seconds 
```

I uninstall old python 2.7.* and now with python 3.7.1 all ok.

**Comment by xC0000005** [comment]
I have the Lerdge X board back in a real printer, so that’s progress. FRAM is working now on S and X, I’ll submit yet another iteration in my quest to fix the TRH/MAX issue, and then it’s time for printing, printing, and more printing. X and S will both be relatively stable, the K board is considerably more complex and will take more time. I created a branch that combines the TFT support, this PR,a nd the STM32 upgrade for Lerdge folks to test. Will update this branch as I pull fixes out of there.

**Comment by mapfde** [comment]
Using the more complete branch of @xC0000005 I have my Lerdge-S now in a printer that is being built. 

Tested and working:
- motor movement (4 + 1 on expansion)
- endstops (6)
- fans (2 + 1 on expansion)
- heaters (2 + 1 on expansion)
- thermistors (which need one pin low and another one high to work) (2 + 1)
- tft, encoder, touchscreen
- rgb led strip (on expansion)
- sdcard access
- settings stored in i2c FRAM
- 3dtouch (servo + switch plug)

Not tested:
- thermocouples (do not have one)
- actual printing (need to fix endstops properly)

Open problems:
- why e1 is turning in opposite direction to z when used as z2 (most likely hardware, as e1 is on an expansion board)

**Comment by jmz52** [comment]
>     * why e1 is turning in opposite direction to z when used as z2 (most likely hardware, as e1 is on an expansion board)

Rotate cable from expansion board to motor by 180 degrees.

**Comment by Evg33** [comment]
J.C. Nelson: Congrats ! :)

**Review by ellensp — COMMENTED** [review]
This commit undid https://github.com/MarlinFirmware/Marlin/pull/18327/files This breaks USB_FLASH_DRIVE_SUPPORT on lpc1768/9 
Need to rename USB_DEVICE_DESCRIPTOR back to USB_FD_DEVICE_DESCRIPTOR and USB_ENDPOINT_DESCRIPTOR to USB_FD_ENDPOINT_DESCRIPTOR and USB_CONFIGURATION_DESCRIPTOR to USB_FD_CONFIGURATION_DESCRIPTOR and USB_INTERFACE_DESCRIPTOR to USB_FD_INTERFACE_DESCRIPTOR

**Comment by cjsoong** [comment]
@xC0000005 
PS_ON_PIN of LERDGE-X is not defined, so AUTO POWER OFF cannot be used. Will it be added later?

**Comment by xC0000005** [comment]
For LerdgeX, I believe this is PC15. My Lerdge power module is somewhere between Taiwan and the US and has been for weeks, so I can’t actually test. It looks like the same pin is both power sense and auto-off (I’m not sure how this works) but you can try defining the pin to PC15 and see if it works. Once I have a module and can test I’ll sort it out.

**Comment by cjsoong** [comment]
Pin_PC15 is used in E0_FAN on my machine, so PS_on_pin should have another pin define.

**Comment by cjsoong** [comment]
@xC0000005 
The result of my test is that the ps_on_pin of Lerdge-x should be PC14.

**Comment by xC0000005** [comment]
I will update this. Lerdge claims this works for both power off and power loss sense. Have you tested power loss sense?

**Comment by cjsoong** [comment]
@xC0000005 
The test has these problems: 
1. Select the option that was not confirmed when printing the file.  
2. LED PIN PC7 has no effect.  
3. POWER LOSS PIN PC14 is unstable. Sometimes an outage message will appear when Power is off, but there is no printer recovery on power on.  
4. The screen for enabling touch function will jump randomly.

**Comment by jmz52** [comment]
@cjsoong 
Please provide details for
- Select the option that was not confirmed when printing the file.
- The screen for enabling touch function will jump randomly.

**Comment by cjsoong** [comment]
@jmz52 
There should be a confirmation option in some operations, but it does not appear, refer to the picture.  
After enabling touch, the LCD menu is very unstable and will automatically jump to other options, and the number will automatically increase or decrease.
IMG_20200729_094822

> @cjsoong
> Please provide details for
> 
> * Select the option that was not confirmed when printing the file.
> * The screen for enabling touch function will jump randomly.

**Comment by jmz52** [comment]
@cjsoong
Please provide link to where you downloaded code from.
Do you have rotary (optional) encoder? If you do, please check if menu operates correctly when using encoder with TOUCH_SCREEN option disabled.
Is Marlin responding to COM port when you have EEPROM init screen with no buttons?

**Review by nikdfzz — APPROVED** [review]

**Pull request merged**
